File: src/routes/migrate.ts
import { Router, Request, Response } from "express";

import { asyncHandler } from "../middleware/asyncHandler";
import { migrateCalibre } from "../lib/migrate/calibre";
import { ApiError } from "../types/api";
import { checkAuth, checkRoles } from "../lib/auth";

export const migrateRouter = Router();

migrateRouter.post(
  "/migrate/calibre",
  checkAuth,
  asyncHandler(async (req: Request, res: Response) => {
    await checkRoles(req.headers.user_roles as string, "manage_library");

    const { calibrePath, libraryName, libraryMetadataProvider } = req.body;

    if (!calibrePath) {
      throw new ApiError(400, "Calibre path is required");
    }

    if (!libraryName) {
      throw new ApiError(400, "Library name is required");
    }

    if (!libraryMetadataProvider) {
      throw new ApiError(400, "Library metadata provider is required");
    }

    migrateCalibre(
      calibrePath as string,
      libraryName as string,
      libraryMetadataProvider as string,
      req.headers.user_id ? Number(req.headers.user_id) : 0
    );
    res.json({ status: true });
  })
);

export default migrateRouter;

File: internal/api/handlers/migrate.go
package handlers

import (
	"log"
	"net/http"

	"github.com/gin-gonic/gin"

	"github.com/devourer/server/internal/auth"
	"github.com/devourer/server/internal/migrate"
)

// MigrateCalibre handles POST /migrate/calibre
func (h *Handlers) MigrateCalibre(c *gin.Context) {
	var body struct {
		CalibrePath             string `json:"calibrePath"`
		LibraryName             string `json:"libraryName"`
		LibraryMetadataProvider string `json:"libraryMetadataProvider"`
	}
	if err := c.ShouldBindJSON(&body); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"status": false, "message": "Invalid request body"})
		return
	}
	if body.CalibrePath == "" {
		c.JSON(http.StatusBadRequest, gin.H{"status": false, "message": "Calibre path is required"})
		return
	}
	if body.LibraryName == "" {
		c.JSON(http.StatusBadRequest, gin.H{"status": false, "message": "Library name is required"})
		return
	}
	if body.LibraryMetadataProvider == "" {
		c.JSON(http.StatusBadRequest, gin.H{"status": false, "message": "Library metadata provider is required"})
		return
	}

	userID, _ := c.Get(auth.CtxUserID)
	uid, _ := userID.(int64)

	d := h.DB
	go func() {
		if err := migrate.MigrateCalibre(d, body.CalibrePath, body.LibraryName, body.LibraryMetadataProvider, uid); err != nil {
			log.Printf("[Migrate] Calibre migration failed: %v", err)
		} else {
			log.Printf("[Migrate] Calibre migration completed for library %q", body.LibraryName)
		}
	}()

	c.JSON(http.StatusOK, gin.H{"status": true, "message": "Calibre migration started"})
}
